pkg/storage/stores/shipper/storage/cached_client_test.go

```diff
@@ -0,0 +1,183 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/loki/pkg/storage/chunk"
+)
+
+type mockObjectClient struct {
+	chunk.ObjectClient
+	storageObjects []chunk.StorageObject
+	errResp        error
+	listCallsCount int
+	listDelay      time.Duration
+}
+
+func newMockObjectClient(objects []string) *mockObjectClient {
+	storageObjects := make([]chunk.StorageObject, 0, len(objects))
+	for _, objectName := range objects {
+		storageObjects = append(storageObjects, chunk.StorageObject{
+			Key: objectName,
+		})
+	}
+
+	return &mockObjectClient{
+		storageObjects: storageObjects,
+	}
+}
+
+func (m *mockObjectClient) List(_ context.Context, _, _ string) ([]chunk.StorageObject, []chunk.StorageCommonPrefix, error) {
+	defer func() {
+		time.Sleep(m.listDelay)
+		m.listCallsCount++
+	}()
+
+	if m.errResp != nil {
+		return nil, nil, m.errResp
+	}
+
+	return m.storageObjects, []chunk.StorageCommonPrefix{}, nil
+}
+
+func TestCachedObjectClient(t *testing.T) {
+	objectsInStorage := []string{
+		// table with just common dbs
+		"table1/db1.gz",
+		"table1/db2.gz",
+
+		// table with both common and user dbs
+		"table2/db1.gz",
+		"table2/user1/db1.gz",
+
+		// table with just user dbs
+		"table3/user1/db1.gz",
+		"table3/user1/db2.gz",
+	}
+
+	objectClient := newMockObjectClient(objectsInStorage)
+	cachedObjectClient := newCachedObjectClient(objectClient)
+
+	// list tables
+	objects, commonPrefixes, err := cachedObjectClient.List(context.Background(), "", "")
+	require.NoError(t, err)
+	require.Equal(t, 1, objectClient.listCallsCount)
+	require.Equal(t, objects, []chunk.StorageObject{})
```

Comment:
Arguments of the Equal function are in "incorrect" order:

```suggestion
	require.Equal(t, []chunk.StorageObject{}, objects)
```

The function interface is

```
func Equal(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})
```

This isn't a problem as long as `expected` and `actual` are equal, but the test error message is misleading in case they aren't.

Reply:
Guess this is not only a problem in your test, but we have that all over the place.

Reply:
yeah, sorry I messed up the order. Fixed it.